lib/galaxy/workflow/modules.py

```diff
@@ -103,6 +105,99 @@ def __str__(self):
 NO_REPLACEMENT = NoReplacement()
 
 
+class ConditionalStepWhen(BooleanToolParameter):
+    pass
+
+
+def evaluate_value_from_expressions(progress, step, execution_state, extra_step_state):
+    when_expression = step.when_expression
+    value_from_expressions = {}
+
+    if execution_state:
+        for key in execution_state.inputs.keys():
+            step_input = step.inputs_by_name.get(key)
+            if step_input and step_input.value_from is not None:
+                value_from_expressions[key] = step_input.value_from
+
+    if not value_from_expressions and when_expression is None:
+        return {}
+
+    hda_references = []
+
+    def to_cwl(value):
```

Comment:
This is so complicated and we're only basically testing boolean parameters right?

What if an integer is sent - what happens? A dataset? Etc...

Can we have a test that the workflow dies if an invalid expression is set in there?

Reply:
I've included a bunch of tests and I've also added "invocation messages", see the screenshot in the description ...  I think we're now displaying (almost?) all of the possible workflow errors in the API and UI.